pytoda/datasets/set_matching_dataset.py

```diff
@@ -0,0 +1,297 @@
+        """
+
+        set_1 = self.datasets[0][index]
+        set_2 = self.get_set_2_element(set_1, index)
+        set_1, set_2 = self.crop_set_lengths(set_1, set_2, index)
+        targets_12, targets_21 = self.get_targets(set_1, set_2)
+        return set_1, set_2, targets_12, targets_21
+
+
+class CollatorSetMatching:
+    """Contains function to pad data returned by dataloader."""
+
+    def __init__(
+        self,
+        dim: int,
+        max_length: int,
+        batch_first: bool = True,
+        device: torch.device = (
+            torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+        ),
+    ):
+        """Constructor.
+
+        Args:
+            dim (int): Dimension of the data.
+            max_length (int): Maximum set length.
+            batch_first (bool, optional): Whether batch size is the first
+                dimension or not. Defaults to True.
+            device (torch.device): device where the tensors are stored.
+                Defaults to gpu, if available.
+        """
+        super(CollatorSetMatching, self).__init__()
+        self.dim = dim
+        self.max_len = max_length
+        self.batch_first = eval(batch_first)
+        self.device = device
+
+    def __call__(self, DataLoaderBatch: Tuple) -> Tuple:
+        """Collate function for batch-wise padding of samples.
+
+        Args:
+            DataLoaderBatch (Tuple): Tuple of tensors returned by get_item of the
+            dataset class.
+
+        Returns:
+            Tuple: Tuple of padded input tensors and tensor of set lengths.
+        """
+
+        batch_size = len(DataLoaderBatch)
+        batch_split = list(zip(*DataLoaderBatch))
+
+        sets1, sets2, targs12, targs21 = (
+            batch_split[0],
+            batch_split[1],
+            batch_split[2],
+            batch_split[3],
+        )
+
+        lengths = list(map(len, sets1))
+        pad_token = 10.0
+
+        padded_sets1 = torch.full(
+            (batch_size, self.max_len, self.dim), pad_token, device=self.device
+        )
+        padded_sets2 = torch.full(
+            (batch_size, self.max_len, self.dim), pad_token, device=self.device
+        )
+        targets12 = np.tile(np.arange(self.max_len), (batch_size, 1))
+        targets21 = np.tile(np.arange(self.max_len), (batch_size, 1))
+        targets12 = torch.from_numpy(targets12).to(self.device)
+        targets21 = torch.from_numpy(targets21).to(self.device)
+
+        for i, l in enumerate(lengths):
+            padded_sets1[i, 0:l, :] = sets1[i][0:l, :]
+            padded_sets2[i, 0:l, :] = sets2[i][0:l, :]
+
+            targets12[i, 0:l] = targs12[i][:]
+            targets21[i, 0:l] = targs21[i][:]
+
+        if self.batch_first is False:
+            padded_sets1, padded_sets2 = (
+                padded_sets1.permute(1, 0, 2),
+                padded_sets2.permute(1, 0, 2),
+            )
+
+        return padded_sets1, padded_sets2, targets12, targets21, torch.tensor(lengths)
```

Comment:
I agree that padding is a generic task, but wrt to set matching, the nature of the targets warrants the need for it to be slightly customised. 1. The lengths are always fixed to a certain max length since probabilities are returned and subsequently used in loss calculation, but I see there is provision for it be to replaced by self.max_length. 2. The targets are a sequence of integers ordered according to how the elements are matched. So the padding token for targets are not what is provided in __init__ but rather the sequence of the remaining integers.

Reply:
Ok that's a good idea!! But I can't think of a way to split the batch in a generic way. Like in my case for ex, there are 2 "features" and 2 "targets". Usually you just have features and targets, right?

Reply:
so will need an additional argument that specifies how many features and targets exist.

Reply:
My idea was to split the batch in the wrapper in a very specific known way.

But the solution you came up with doing different padding in the same collator call also works. In this case the wrapper again only adds lengths.